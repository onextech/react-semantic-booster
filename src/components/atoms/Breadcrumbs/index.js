import React from 'react';
import PropTypes from 'prop-types';
import { Link, withRouter } from 'react-router-dom';
import { Breadcrumb } from 'semantic-ui-react';
import { capitalizeFirstLetter } from '@onextech/react-apollo-utils';


const Breadcrumbs = ({ match, customCrumbs }) => {
  const crumbs = match.url.split('/');

  /**
   * Set custom routes on certain params
   * @param {string} crumb
   * @param {number} index
   * @param {{}} rules - set of keys (match.params path) and values (route to go to)
   * @returns {string}
   */
  const getCrumbSlug = (crumb, index, rules) => {
    Object.entries(rules).forEach(([k, v]) => {
      if (k === crumb) {
        return v;
      }
      return crumbs.slice(0, index + 1).join('/');
    });
    return '';
  };

  return (
    <Breadcrumb style={{ margin: '0.75em auto' }}>
      {
        crumbs.map((crumb, index) => {
          const slug = getCrumbSlug(crumb, index, customCrumbs);

          const jsx = [
            <Breadcrumb.Section key={crumb}>
              <Link to={slug}>{index === 0 ? 'Home' : capitalizeFirstLetter(crumb)}</Link>
            </Breadcrumb.Section>,
          ];

          if (index !== crumbs.length - 1) {
            jsx.push(<Breadcrumb.Divider key={`${crumb}-divider`} icon="right angle" />);
          }

          return jsx;
        })
      }
    </Breadcrumb>
  );
};

Breadcrumbs.propTypes = {
  match: PropTypes.object.isRequired,
  customCrumbs: PropTypes.object,
};

Breadcrumbs.defaultProps = {
  customCrumbs: {},
};

export default withRouter(Breadcrumbs);
