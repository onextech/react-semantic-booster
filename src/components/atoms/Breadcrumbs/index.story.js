import React from 'react';
import { storiesOf } from '@storybook/react';
import Breadcrumbs from '.';


storiesOf('Breadcrumbs', module)
  .add('Default', () => (
    <Breadcrumbs />
  ));
