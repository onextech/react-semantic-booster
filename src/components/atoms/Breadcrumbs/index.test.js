import React from 'react';
import renderer from 'react-test-renderer';
import { render } from 'enzyme';
import { MemoryRouter as Router } from 'react-router-dom';
import Breadcrumbs from '.';


describe('<Breadcrumbs>', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Router><Breadcrumbs /></Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('is made of semantic-ui', () => {
    const wrapper = render(<Router><Breadcrumbs>Hello World</Breadcrumbs></Router>);
    expect(wrapper.hasClass('ui breadcrumb')).toBe(true);
  });

  it('is always made up by at least three children by default', () => {
    const wrapper = render(<Router><Breadcrumbs>Hello World</Breadcrumbs></Router>);
    expect(wrapper.children().length).toBe(3);
  });

  it('has at least a link inside', () => {
    const wrapper = render(<Router><Breadcrumbs>Hello World</Breadcrumbs></Router>);
    expect(wrapper.find('a').length).toBeGreaterThanOrEqual(1);
  });
});
