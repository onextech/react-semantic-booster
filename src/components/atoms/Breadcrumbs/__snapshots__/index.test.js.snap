// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs> renders correctly 1`] = `
<div
  className="ui breadcrumb"
  style={
    Object {
      "margin": "0.75em auto",
    }
  }
>
  <div
    className="section"
    href={undefined}
    onClick={[Function]}
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
  </div>
  <i
    aria-hidden="true"
    className="right angle icon divider"
  />
  <div
    className="section"
    href={undefined}
    onClick={[Function]}
  >
    <a
      href="/"
      onClick={[Function]}
    >
      
    </a>
  </div>
</div>
`;
